### PR TITLE
[GEN] Change existing OCL lowering to the new interface

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Please do not squash and merge this PR.

### CODE REVIEW

**Comment by whitneywhtsang** [comment]
Awaiting for a release driver that contains the support of the OCL builtin, but ready for review.

**Review by chengjunlu — APPROVED** [review]

**Review by victor-eds — COMMENTED** [review]
Are CI failures related to these changes?

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 1]
```diff
@@ -248,30 +248,36 @@ static bool isOCLBuiltinAvailable(TritonGEN::Matrix2DBlockLoadOp op) {
...
-createGenISA2DBlockRead(TritonGEN::Matrix2DBlockLoadOp op,
-                        ConversionPatternRewriter &rewriter) {
+static Value createGenISA2DBlockRead(TritonGEN::Matrix2DBlockLoadOp op,
+                                     ConversionPatternRewriter &rewriter) {
   MLIRContext *context = rewriter.getContext();
-  Type resType = op->getResultTypes()[0];
+  VectorType resType = op.getRes().getType();
   Location loc = op->getLoc();
 
   // FIXME: Use the OpenCL API also for all other variants.
```
Comment (FMarno) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:257]: remove fixme comment

File: test/TritonGEN/tritongen-to-llvm.mlir [hunk 2]
```diff
@@ -367,16 +367,19 @@ llvm.func @triton_gen.dpas.f32(%c : vector<8xf32>, %a : vector<4xf32>, %b : vect
 
 // -----
 
-// CHECK: llvm.func spir_funccc @intel_subgroup_block_read_u8_m8k32v2(!llvm.ptr<1>, i32, i32, i32, vector<2xi32>) -> vector<16xi16> attributes {passthrough = ["convergent"]}
+// CHECK: llvm.func spir_funccc @_Z40intel_sub_group_2d_block_read_8b_8r32x2cPU3AS1viiiDv2_iPt(!llvm.ptr<1>, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["convergent"]}
```
Comment (FMarno) [inline test/TritonGEN/tritongen-to-llvm.mlir:370]: I think a test for all the other variations would be helpful.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp [hunk 3]
```diff
@@ -248,30 +248,36 @@ static bool isOCLBuiltinAvailable(TritonGEN::Matrix2DBlockLoadOp op) {
...
-    return createDeviceFunctionCall(rewriter, fnName, resType, argTypes, args,
-                                    true /*convergent*/);
+    SmallVector<Type> argTypes{
+        ptr_ty(context, 1), i32_ty, i32_ty, i32_ty, vecType, ptr_ty(context)};
+    SmallVector<Value> args{op.getPtr(),        op.getBaseWidth(),
+                            op.getBaseHeight(), op.getBasePitch(),
+                            byteCoord,          dest};
+    createDeviceFunctionCall(rewriter, fnName, void_ty(context), argTypes, args,
+                             true /*convergent*/);
+    return rewriter.create<LLVM::LoadOp>(loc, resType, dest);
```
Comment (etiotto) [inline third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:280]: Have we verified that the new OCL builtins do not cause performance degradation given that now we have to allocate/call/load  in loops ?

Comment (etiotto) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:280]: If performance isn't on par with the best we have so far then we need to know before merging this PR.

**Comment by whitneywhtsang** [comment]
> Are CI failures related to these changes?

Yes, will pass when we update to the next agama rolling driver.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 1 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:257]: Cannot remove yet, as we are still using GenISA for some variants.

File: test/TritonGEN/tritongen-to-llvm.mlir (hunk 2 above)
Comment (whitneywhtsang) [reply test/TritonGEN/tritongen-to-llvm.mlir:370]: Added in https://github.com/intel/intel-xpu-backend-for-triton/pull/1259.

**Review by FMarno — APPROVED** [review]
Code seems fine to me

File: test/TritonGEN/tritongen-to-llvm.mlir [hunk 4]
```diff
@@ -297,16 +297,19 @@ llvm.func @triton_gen.dpas.f32(%c : vector<8xf32>, %a : vector<4xf32>, %b : vect
 
 // -----
 
-// CHECK: llvm.func spir_funccc @intel_subgroup_block_read_u8_m8k32v2(!llvm.ptr<1> {llvm.nonnull}, i32, i32, i32, vector<2xi32>) -> vector<16xi16> attributes {passthrough = ["nounwind"]}
+// CHECK: llvm.func spir_funccc @_Z40intel_sub_group_2d_block_read_8b_8r32x2cPU3AS1viiiDv2_iPt(!llvm.ptr<1> {llvm.nonnull}, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["nounwind"]}
 
 llvm.func @triton_gen.2Dblockload(%ptr : !llvm.ptr<1>, %base_width : i32, %base_height : i32, %base_pitch : i32, %x : i32, %y : i32) {
   // CHECK:     llvm.func @triton_gen.2Dblockload(%arg0: !llvm.ptr<1>, %arg1: i32, %arg2: i32, %arg3: i32, %arg4: i32, %arg5: i32) {
+  // CHECK:  [[SIXTY:%.*]] = llvm.mlir.constant(16 : i32) : i32
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/TritonGEN/tritongen-to-llvm.mlir]: done.

File: test/TritonGEN/tritongen-to-llvm.mlir [hunk 5]
```diff
@@ -297,16 +297,19 @@ llvm.func @triton_gen.dpas.f32(%c : vector<8xf32>, %a : vector<4xf32>, %b : vect
 
 // -----
 
-// CHECK: llvm.func spir_funccc @intel_subgroup_block_read_u8_m8k32v2(!llvm.ptr<1> {llvm.nonnull}, i32, i32, i32, vector<2xi32>) -> vector<16xi16> attributes {passthrough = ["nounwind"]}
+// CHECK: llvm.func spir_funccc @_Z40intel_sub_group_2d_block_read_8b_8r32x2cPU3AS1viiiDv2_iPt(!llvm.ptr<1> {llvm.nonnull}, i32, i32, i32, vector<2xi32>, !llvm.ptr) attributes {passthrough = ["nounwind"]}
```
[the remark this answers is not in the document]
Comment (whitneywhtsang) [reply test/TritonGEN/tritongen-to-llvm.mlir]: done.

File: third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonGENToLLVM/TritonGENToLLVMPass.cpp:316]: Measured no performance impact.

**Review by etiotto — APPROVED** [review]

**Pull request merged**